src/coreclr/nativeaot/Runtime/eventpipe/ep-rt-aot.cpp

```diff
@@ -88,9 +95,57 @@ ep_rt_aot_sample_profiler_write_sampling_event_for_threads (
 const ep_char8_t *
 ep_rt_aot_entrypoint_assembly_name_get_utf8 (void) 
 {
-    // shipping criteria: no EVENTPIPE-NATIVEAOT-TODO left in the codebase
-    // TODO: Implement EventPipe assembly name - return filename in nativeaot?
-    return reinterpret_cast<const ep_char8_t*>("");
+    // Cannot use __cpp_threadsafe_static_init feature since it will bring in the C++ runtime and need to use threadsafe way to initialize entrypoint_assembly_name
+    static const ep_char8_t * entrypoint_assembly_name = nullptr;
+    if (entrypoint_assembly_name == nullptr) {
+        ep_char8_t * entrypoint_assembly_name_local;
+        const TCHAR * wszModuleFileName = NULL;
+        HANDLE moduleHandle = PalGetModuleHandleFromPointer((void*)&ep_rt_aot_entrypoint_assembly_name_get_utf8);
+        if(PalGetModuleFileName(&wszModuleFileName, moduleHandle) == 0) {
+            entrypoint_assembly_name_local = reinterpret_cast<ep_char8_t *>(malloc(1));
+            if(entrypoint_assembly_name_local==NULL) {
+                return NULL;
+            }
+            *entrypoint_assembly_name_local = '\0';
+        }
+        else {
+#ifdef HOST_WINDOWS
+            const wchar_t* process_name_const = wcsrchr(wszModuleFileName, DIRECTORY_SEPARATOR_CHAR);
+            if (process_name_const != NULL) {
+                process_name_const++;
+            }
+            size_t len = -1;
+            const wchar_t* extension = wcsrchr(process_name_const, '.');
+            if (extension != NULL) {
+                len = extension - process_name_const;
+            }
+            const ep_char16_t* process_name = reinterpret_cast<const ep_char16_t *>(process_name_const);
+            entrypoint_assembly_name_local = ep_rt_utf16_to_utf8_string(process_name, len);
+#else
+            const ep_char8_t* process_name_const = strrchr(wszModuleFileName, DIRECTORY_SEPARATOR_CHAR);
+            if (process_name_const != NULL) {
+                process_name_const++;
+            }
+            size_t len = strlen(process_name_const);
+            const ep_char8_t *extension = strrchr(process_name_const, '.');
+            if (extension != NULL) {
+                len = len - strlen(extension);
```

Comment:
Why is this length computed differently from the Windows section above? The same logic should work for both?